src/RoleSectionStorage.php

```diff
@@ -39,41 +39,88 @@ class RoleSectionStorage implements RoleSectionStorageInterface {
    */
   public function __construct(StateInterface $state, EntityTypeManagerInterface $entityTypeManager) {
     $this->state = $state;
+    $this->entityTypeManager = $entityTypeManager;
     $this->roleStorage = $entityTypeManager->getStorage('user_role');
   }
 
+  /**
+   * \Drupal\Core\Entity\EntityStorageInterface
+   */
+  protected function sectionStorage() {
+    // The entity build process takes place too early in the call stack and we
+    // have test fails if we add this to the __construct().
+    return $this->entityTypeManager->getStorage('section_association');
+  }
+
   /**
    * {@inheritdoc}
+   *
+   * @TODO: refactor.
    */
   public function addRole(AccessSchemeInterface $scheme, $role_id, array $sections = []) {
-    $settings = $this->loadRoleSections($scheme, $role_id);
     foreach ($sections as $id) {
-      $settings[$id] = 1;
+      // @TODO: This is tortured logic and probably much easier to handle.
+      if ($section_association = $this->sectionStorage()->loadSection($scheme->id(), $id)) {
+        $mew_values = [];
+        if ($values = $section_association->get('role_id')) {
+          foreach ($values as $delta => $value) {
+            $target = $value->getValue();
+            $new_values[] = $target['target_id'];
+          }
+          $new_values[] = $role_id;
+          $section_association->set('role_id', array_unique($new_values));
+        }
+        else {
+          $section_association->set('role_id', [$role_id]);
+        }
+        $section_association->setNewRevision();
+      }
+      else {
+        $values = [
+          'access_scheme' => $scheme->id(),
+          'section_id' => $id,
+          'role_id' => [$role_id],
+        ];
+        $section_association = $this->sectionStorage()->create($values);
+      }
+      $section_association->save();
     }
-    $this->saveRoleSections($scheme, $role_id, $settings);
   }
 
   /**
    * {@inheritdoc}
+   *
+   * @TODO: refactor.
    */
   public function removeRole(AccessSchemeInterface $scheme, $role_id, array $sections = []) {
-    $settings = $this->loadRoleSections($scheme, $role_id);
     foreach ($sections as $id) {
-      if (isset($settings[$id])) {
-        unset($settings[$id]);
+      // @TODO: This is tortured logic and probably much easier to handle.
+      if ($section_association = $this->sectionStorage()->loadSection($scheme->id(), $id)) {
+        $new_values = [];
+        if ($values = $section_association->get('role_id')) {
```

Comment:
should be able to use array_filter and array_column to simplify this

Reply:
If I can make it readable. It's a nice cleanup for later.

My suspicion is that the add/remove methods can be abstracted into the SectionAssociationStorage and written once.